### ISSUE TITLE
More ideas for tests

### ISSUE THREAD

**Issue by kevinlyles** [rsnapshot/rsnapshot#60]
* file local vs absolute/link/directory traversal checks for:
	* -c?
	* include_conf (with args?)
	* cmd_rsync (no args)
	* cmd_ssh (no args)
	* cmd_cp (no args)
	* cmd_rm (no args)
	* cmd_logger (with args?)
	* cmd_du (no args)
	* cmd_rsnapshot_diff (no args)
	* cmd_preexec (with args?)
	* cmd_postexec (with args?)
	* linux_lvm_cmd_lvcreate (no args, backup level too)
	* linux_lvm_cmd_lvremove (no args, backup level too)
	* linux_lvm_cmd_mount (no args, backup level too)
	* linux_lvm_cmd_umount (no args, backup level too)
	* logfile
	* include_file (backup level too?)
	* exclude_file (backup level too?)
	* backup
	* backup_script
* Mixing retain and interval
* Mixing log methods
* Check command line options:
	* -t: verify that no changes are made for a series of commands
	* -c: probably covered by all of the other tests
	* -x: specify a custom "rsync" and check for -x with it
	* -v, -q, -V, -D: unsure
	* Check multiple retains
* Test lockfiles hanging around (on every test?)

**Comment by bebehei** [comment]
:+1: This looks great. Right now, every test is enhancing rsnapshots' test suite. We have 7 tests against 4000 SLOCs in `rsnapshot-program.pl`.

But I have one single objection, before we go on writing tests. Right now, there is a unittest at `t/*.t` and a config-file at `t/support/etc/*.conf` and `t/support/files`. On the long term, we will have loads of tests and I believe, we will have problems later keeping them together without loosing the overview.

bf7cfac78 tackles this problem, but I don't think the structure of our test-directory is right.

I would like to have a structure like this

    t/
    |- $testname/
    |-----------/test.t.in
    \-----------/rsnapshot.conf.in

cc @sam-at-github

**Comment by sam-at-github** [comment]
@bebehei  :+1: Absolutely.

**Comment by sam-at-github** [comment]
@kevinlyles Great list!

**Comment by bebehei** [comment]
Bit OT: there is nothing about test-cases added in CONTRIBUTING.md. We have to port `NOTES/add_test_case`, adapt it and maybe rewrite parts. At the end we have to put it into CONTRIBUTING.md

**Comment by bebehei** [comment]
Ok, guys. I think we have to think about, what to test and how to test the whole bunch of things. I've stumbled over [a stackoverflow-thread mentioning modulinos](http://stackoverflow.com/questions/9134408/is-perl-unit-testing-only-for-modules-not-programs).

What do you think about this?

I bet that using the output of rsync and the exit-code gives enough room for positive falses.

**Comment by DrHyde** [comment]
Breaking the monolithic application into separate modules would certainly help to make it easier to test. However, the very act of breaking it up potentially introduces bugs. I'm lairy of doing that without having better test coverage than we've already got, which is a bit of a circular problem. [Here's a report](http://www.cantrell.org.uk/david/private/rsnapshot-coverage/coverage.html) on our current test coverage, generated thus:

    $ PERL5OPT=-MDevel::Cover make test
    $ cover

**Comment by bebehei** [comment]
You're right. This might decrease the test-coverate at the beginning, but I don't mind this in first place.

Right now, my general interest is to search for a good solution to write tests in the long-term-run. This has to be a reliable solution, and it should assure that the program gets checked in its units and in its entire setup. Additionally, to write a test has to be done with small effort.

If we just start writing tests without having a concept, we'll end up having more trouble than we have right now. The test-suite used right now, does not satisfy me. I believe there is a better solution than the one right now used.

**Comment by DrHyde** [comment]
Even if we rip the guts apart and change everything to make it more testable, we wouldn't want to change the public interface - that is, the command line and config file format - that we've currently got, so tests like what we've already got are still useful. They are at least useful regression tests even if they're inefficient and not particularly useful for debugging.

**Comment by bebehei** [comment]
I don't want to change the public interface either. Changing the public interface is as I imagine modulinos not neccessary.

> They are at least useful regression tests even if they're inefficient and not particularly useful for debugging.

Yes, they are useful. But these tests are not satisfying me. They're not efficient. I want to improve the test-suite to have full test-coverage. That requires more quantity and efficiency. And before someone starts with this, we should have a basic layout of the tests.

For example, this is a referenced check: `gnu_cp.t`

    ok(! `/usr/bin/perl ./rsnapshot -c t/support/etc/gnu_cp.conf hourly`);

What does it actually do? It counts the lines of the output. No exit-code-check at all. Besides that the check is right now unfailable, I think just checking if there are lines printed out, is not the way how to successfully do a test.

I pledge to use another technique in our test-suite. I want to have a more specific tests. I haven't decided yet which one to vote for, and probably there're plenty of other techniques instead of modulinos.

**Comment by sam-at-github** [comment]
> That report will go away the next time I make a change to my website - is there somewhere better to put things like this? Note that while Coveralls.io is awesome ..

In the README.md like these?: https://github.com/markdalgleish/gulp-coveralls, https://github.com/kt3k/coveralls-gradle-plugin

> Additionally, to write a test has to be done with small effort.

:+1: yep.

Not in favor of modulularizing, too much effort for no gain. I'm OK with the current setup. Just need some minor restructuring, an maybe some guidelines to follow for new tests.

**Comment by DrHyde** [comment]
Re fiddling with configure.ac, I'm very strongly in favour of packaging "cpan-style", so the incantation to install from source becomes 'perl Makefile.PL && make test && sudo make install'. That will automatically generate man-pages and fiddle with #! lines to point at an appropriate build of perl, and find all the .t files.

**Comment by bebehei** [comment]
--with-question='Do Makefile.PL or Build.PL have support for common options, like setting another rsync-path?'

**Comment by sam-at-github** [comment]
Definitely worth a look. I;m in favor of that too  so long as it supports those configuration options. I'm not sure, but it should. 

Also note I just pushed a branch to rsnapshot. fix/tests. This closes ~3 of my earlier tickets on tests. I'm still working on a autoconf issue, but have a fix for I'm working on. I realize the way a this conversion is going be change the build and/or test setup soon, but I think its worth having this done sooner rather than later, and they are not that major.

**Comment by bebehei** [comment]
@sam-at-github Had you been able to work on the testsuite? I'm waiting for you, to merge #74.

In #74 there are tests for `cmd_preexec` and `cmd_postexec` added.

**Comment by bebehei** [comment]
`-c` will get testsuite-support at #77. The tests are not added there yet, but they will follow.

**Comment by sam-at-github** [comment]
> @sam-at-github Had you been able to work on the testsuite? I'm waiting for you, to merge #74.
Merged, but haven't been able to work on test suite. Pretty busy. Will try to get to it though.
